src/btrfs.cpp

```diff
@@ -0,0 +1,154 @@
+** and appearing in the file license.lgpl included in the packaging
+** of this file.
+**
+** This library is free software; you can redistribute it and/or
+** modify it under the terms of the GNU Lesser General Public
+** License version 2.1 as published by the Free Software Foundation
+** and appearing in the file license.lgpl included in the packaging
+** of this file.
+**
+** This library is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+** Lesser General Public License for more details.
+**
+****************************************************************************************/
+
+
+#include "btrfs.h"
+
+#include <QLocale>
+#include <QRegExp>
+#include <QDebug>
+
+namespace
+{
+// path of the btrfs tool
+const QString BTRFS_PATH("/usr/sbin/btrfs");
+
+// default root device to use
+const QString ROOT_DEVICE("/dev/mmcblk0p28");
+
+// regexp to retrieve size usage information from btrfs output
+const QRegExp RE_USAGE("size ([0-9.]+\\w+) used ([0-9.]+\\w+)");
+
+qint64 parseSize(const QString& size)
+{
+    double value = -1;
+
+    QLocale defaultLocale = QLocale();
+    QLocale::setDefault(QLocale::C);
+
+    if (size.endsWith("MiB")) {
+        value = size.left(size.length() - 3).toDouble()
+                * 1024 * 1024;
+    } else if (size.endsWith("GiB")) {
+        value = size.left(size.length() - 3).toDouble()
+                * 1024 * 1024 * 1024;
+    } else if (size.endsWith("TiB")) {
+        // one day this may become useful...
+        value = size.left(size.length() - 3).toDouble()
+                * 1024 * 1024 * 1024 * 1024;
+    }
+
+    QLocale::setDefault(defaultLocale);
+    return static_cast<qint64>(value);
+}
+
+}
+
+Btrfs::Btrfs(QObject *parent)
+    : QObject(parent)
+    , m_currentProcess(0)
+{
+
+}
+
+void Btrfs::allocation()
+{
+    if (m_currentProcess) {
+        emit allocationReceived(-1, -1);
+        return;
+    }
+
+    m_currentProcess = new QProcess;
+    m_currentProcess->setProgram(BTRFS_PATH);
+    m_currentProcess->setArguments(QStringList()
+                                   << "filesystem"
+                                   << "show");
+
+    connect(m_currentProcess, SIGNAL(finished(int,QProcess::ExitStatus)),
+            this, SLOT(slotAllocationFinished(int,QProcess::ExitStatus)));
+    m_currentProcess->start(QProcess::ReadOnly);
+}
+
+void Btrfs::balance(int usage)
+{
+    if (m_currentProcess) {
+        emit balanceFinished(false);
+        return;
+    }
+
+    m_currentProcess = new QProcess;
+    m_currentProcess->setProgram(BTRFS_PATH);
+    m_currentProcess->setArguments(QStringList()
+                                   << "balance"
+                                   << "start"
+                                   << QString("-dusage=%1").arg(usage)
+                                   << "/");
```

Comment:
We might need to make that mountpoint configurable as well (in `sailfish-snapshot`, we mount the subvolume-zero in `/sailfish-snapshot/` to operate on it). Or at least make it a named constant, so this single-character path doesn't hide so well in the code if somebody else is looking for it in the future.